### PR TITLE
upstream: fix bug in load stats reporting when local locality is miss…

### PR DESCRIPTION
**Pull request by htuch**
…ing.

Previously, when an EDS assignment had no hosts in the local locality
for a cluster, load reports were not created, since we lost track of
hosts per locality.

In this patch, we now track hosts per locality regardless of the
presence of hosts in the local locality.

Another related change is that we also now bump the
lb_local_cluster_not_ok stat when bypassing zone aware routing due to no
hosts being present in the local locality.

Risk Level: Low
Tests: Additional load stats reporter and zone aware routing tests.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/upstream/eds.cc [hunk 1]
```diff
@@ -108,27 +108,28 @@ void EdsClusterImpl::updateHostsPerLocality(HostSet& host_set, HostVector& new_h
...
+    std::map<Locality, HostVector> hosts_per_locality;
+
+    for (const HostSharedPtr& host : *current_hosts_copy) {
+      hosts_per_locality[Locality(host->locality())].push_back(host);
+    }
 
-      // Populate per_locality hosts only if upstream cluster has hosts in the same locality.
-      if (hosts_per_locality.find(local_locality) != hosts_per_locality.end()) {
-        per_locality.push_back(hosts_per_locality[local_locality]);
+    const bool local = !local_locality.empty() &&
```
Comment (mattklein123) [inline source/common/upstream/eds.cc]: nit: s/local/update_has_local_locality (or something) I think would be clearer.

File: source/common/upstream/eds.cc [hunk 2]
```diff
@@ -108,27 +108,28 @@ void EdsClusterImpl::updateHostsPerLocality(HostSet& host_set, HostVector& new_h
...
+                       hosts_per_locality.find(local_locality) != hosts_per_locality.end();
+
+    if (local) {
```
Comment (mattklein123) [inline source/common/upstream/eds.cc]: There should have been more comments here to begin with. Can you add a comment about how the local locality is first in the vector if it exists?

File: source/common/upstream/load_balancer_impl.cc [hunk 3]
```diff
@@ -228,9 +228,13 @@ void ZoneAwareLoadBalancerBase::resizePerPriorityState() {
 bool ZoneAwareLoadBalancerBase::earlyExitNonLocalityRouting() {
   // We only do locality routing for P=0.
   HostSet& host_set = *priority_set_.hostSetsPerPriority()[0];
+  if (host_set.healthyHostsPerLocality().get().size() < 2) {
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: What's the reasoning behind this change? Can you add more comments?

Comment (htuch) [reply source/common/upstream/load_balancer_impl.cc]: I think this better matches the documented semantics for `lb_local_cluster_not_ok`:

"Local host set *is not set* or it is panic mode for local cluster"

Comment (mattklein123) [reply source/common/upstream/load_balancer_impl.cc]: OK makes sense, thanks.

**Pull request merged**